File: Machine Learning A-Z/Part 1 - Data Preprocessing/data_preprocessing.py
#data_preprocessing.py
import numpy as nu
import matplotlib.pyplot as plt 
import pandas as pd 

#Set the Data.csv folder as the current directory
dataset = pd.read_csv('Data.csv')
#independent variables
#fetch [All the lines, all the columns except last column]
x = dataset.iloc[:, :-1].values
#dependent variables ie the last column- Purchase 
y = dataset.iloc[:, 3].values

#Handing Missing Data: Use Mean of the remaining Data
from sklearn.preprocessing import Imputer
imputer = Imputer(missing_values = 'NaN', strategy = 'mean', axis = 0)
#Fit the mean result into the dataset where the value is NaN 
imputer = imputer.fit(x[:, 1:3])
#replace the missing data 
x[:, 1:3] = imputer.transform(x[:, 1:3])

# Encoding categorical data
# Country and Purchase are categorical data 
# Replace text with numbers since ml is mathematical model, good to replace text with number 
# Dummy encoder: To make sure that the ml model will not consider the order of the numerical values
# create one column for each categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_x = LabelEncoder()
x[:, 0] = labelencoder_x.fit_transform(x[:, 0])
onehotencoder = OneHotEncoder(categorical_features = [0])
x = onehotencoder.fit_transform(x).toarray()
labelencoder_y = LabelEncoder()
y = labelencoder_y.fit_transform(y)

# Split data into Training set and Test set 
from sklearn.cross_validation import train_test_split
# 20% data will be used for training maximum 40%
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)


# ML Model is based on Euclidean Distance 
# Euclidean Distance between P1 and P2 = Sqrt((x2-x1)2 + (y2-y1)2)
#Standardisation x(std) = (x - mean(x))/SD(x)
#Normalization x(norm) = (x - min(x))/(max(x) - min(x))

# Feature Scaling
# Convert data into same range to avoid the impact of large data in one column
from sklearn.preprocessing import StandardScaler
sc_x = StandardScaler()
x_train = sc_x.fit_transform(x_train)
#Train data already fix and so test data not needed to fit againg to the model.
x_test = sc_x.transform(x_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)

File: Machine Learning A-Z/Part 1 - Data Preprocessing/data_preprocessing.R
# Data Preprocessing
# Importing the dataset

#Set Working Directory as the Directory containing Data.csv
dataset = read.csv('Data.csv')

#Missing Data: Use mean of the remaining data

dataset$Age = ifelse(is.na(dataset$Age), 
				ave(dataset$Age, FUN = function(x) mean(x, na.rm = TRUE)), 
				dataset$Age)

dataset$Salary = ifelse(is.na(dataset$Salary), 
				ave(dataset$Salary, FUN = function(x) mean(x, na.rm = TRUE)), 
				dataset$Salary)

# Encoding categorical data
#c is a vector in R
dataset$Country = factor(dataset$Country,
					levels = c('France', 'Spain', 'Germany'),
					labels = c(1, 2, 3))

dataset$Purchased = factor(dataset$Purchased,
					levels = c('No', 'Yes'),
					labels = c(0, 1))

# Split data into Training set and Test set 
# Once installed the package, comment
# install.packages('caTools')
library(caTools)
set.seed(123)
# In Python set for test set , in R we need to set for training set and so 80%
split = sample.split(dataset$Purchased, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature Scaling
# Category columns country and purchased are string values and so needs to exclude from the set to avoid the error.
training_set[, 2:3] = scale(training_set[, 2:3])
test_set[, 2:3] = scale(test_set[, 2:3])

